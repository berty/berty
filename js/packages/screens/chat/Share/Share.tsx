import Clipboard from '@react-native-clipboard/clipboard'
import { useFocusEffect } from '@react-navigation/core'
import { Layout } from '@ui-kitten/components'
import { BarCodeScanner } from 'expo-barcode-scanner'
import { Camera } from 'expo-camera'
import React, { FC, ReactNode, useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { View, Vibration, StatusBar, Share, ScrollView, Platform } from 'react-native'
import { RESULTS } from 'react-native-permissions'
import QRCode from 'react-native-qrcode-svg'

import logo from '@berty/assets/images/1_berty_picto.png'
import ScanTarget from '@berty/assets/logo/scan_target.svg'
import { AccountAvatar } from '@berty/components/avatars'
import { LoaderDots } from '@berty/components/LoaderDots'
import { ButtonSetting, ButtonSettingRow } from '@berty/components/shared-components'
import { UnifiedText } from '@berty/components/shared-components/UnifiedText'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useStyles } from '@berty/contexts/styles'
import { useAccount } from '@berty/hooks'
import { ScreenFC, useNavigation } from '@berty/navigation'
import { useMessengerClient } from '@berty/store'
import { useThemeColor } from '@berty/store/hooks'
import { checkPermissions } from '@berty/utils/react-native/checkPermissions'
import { PermissionType } from '@berty/utils/react-native/permissions'

const QrCode: FC<{ size: number }> = ({ size }) => {
	const client = useMessengerClient()
	const colors = useThemeColor()
	const account = useAccount()
	const [link, setLink] = useState<string>('')

	const getAccountLink = useCallback(async () => {
		if (account.displayName) {
			const ret = await client?.instanceShareableBertyID({
				reset: false,
				displayName: account.displayName,
			})
			if (ret?.internalUrl) {
				setLink(ret?.internalUrl)
			}
		}
	}, [account.displayName, client])

	useEffect(() => {
		getAccountLink()
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [])

	return link ? (
		<QRCode
			size={size}
			value={link}
			logo={logo}
			color={colors['background-header']}
			mode='circle'
			backgroundColor={colors['main-background']}
		/>
	) : (
		<View style={{ width: size, height: size, justifyContent: 'center' }}>
			<LoaderDots />
		</View>
	)
}

const ScanBody: FC<{ visible: boolean }> = ({ visible = true }) => {
	const navigation = useNavigation()
	const { background, margin, flex, column, border } = useStyles()
	const { windowHeight, windowWidth, isGteIpadSize, fontScale } = useAppDimensions()

	const qrScanSize = isGteIpadSize
		? Math.min(windowHeight, windowWidth) * 0.5
		: Math.min(windowHeight * 0.8, windowWidth * 0.8) - 1.25 * (26 * fontScale)

	return (
		<View
			style={[
				background.black,
				margin.horizontal.small,
				column.item.center,
				flex.align.center,
				flex.justify.center,
				border.radius.medium,
				{
					overflow: 'hidden',
					height: qrScanSize,
					aspectRatio: 1,
					position: 'relative',
				},
			]}
		>
			{visible && (
				<Camera
					onBarCodeScanned={({ data, type }) => {
						if (type === BarCodeScanner.Constants.BarCodeType.qr) {
							// I would like to use binary mode in QR but this scanner seems to not support it, extended tests were done
							navigation.navigate('Chat.ManageDeepLink', { type: 'qr', value: data })
							Vibration.vibrate(1000)
						}
					}}
					barCodeScannerSettings={{
						barCodeTypes: [BarCodeScanner.Constants.BarCodeType.qr],
					}}
					style={{
						height: '100%',
						width: '100%',
					}}
				/>
			)}

			<ScanTarget height='75%' width='75%' style={{ position: 'absolute' }} />
		</View>
	)
}

const ShareQr: FC = () => {
	const { margin } = useStyles()
	const { windowWidth, windowHeight, scaleSize } = useAppDimensions()
	const account = useAccount()
	const qrCodeSize = Math.min(windowHeight, windowWidth) * 0.45

	return (
		<View>
			<View
				style={[
					margin.top.big,
					margin.bottom.small,
					{ alignItems: 'center', flexDirection: 'row', justifyContent: 'center' },
				]}
			>
				<View style={[margin.right.small]}>
					<AccountAvatar size={24 * scaleSize} />
				</View>
				<UnifiedText>{account.displayName || ''}</UnifiedText>
			</View>
			<QrCode size={qrCodeSize} />
		</View>
	)
}

const ShareContainer: FC<{ element: ReactNode }> = ({ element, children }) => {
	const colors = useThemeColor()
	const { padding, border, margin } = useStyles()
	const { windowWidth, windowHeight, isGteIpadSize, fontScale } = useAppDimensions()

	const containerSize = isGteIpadSize
		? Math.min(windowHeight, windowWidth) * 0.5
		: Math.min(windowHeight * 0.8, windowWidth * 0.8) - 1.25 * (26 * fontScale)

	return (
		<View
			style={[
				border.radius.bottom.huge,
				padding.bottom.huge,
				padding.top.medium,
				{
					backgroundColor: colors['background-header'],
				},
			]}
		>
			<View
				style={[
					border.radius.medium,
					margin.bottom.big,
					{
						alignSelf: 'center',
						backgroundColor: colors['main-background'],
						alignItems: 'center',
						width: containerSize,
						height: containerSize,
					},
				]}
			>
				{element}
			</View>
			{children}
		</View>
	)
}

export const ShareModal: ScreenFC<'Chat.Share'> = () => {
	const { flex, margin, height, width } = useStyles()
	const colors = useThemeColor()
	const [isScannerVisible, setIsScannerVisible] = useState<boolean>(true)
	const [isScannerSelected, setIsScannerSelected] = useState<boolean>(false)
	const { navigate } = useNavigation()
	const { t } = useTranslation()
	const account = useAccount()
	const url = account.link

	// const status = await checkPermissions(PermissionType.camera, {
	// 	navigate,
	// 	navigateToPermScreenOnProblem: true,
	// })

	const initScanner = useCallback(async () => {
		const status = await checkPermissions(PermissionType.camera)
		setIsScannerSelected(status !== RESULTS.DENIED && status !== RESULTS.BLOCKED)
	}, [])

	useEffect(() => {
		initScanner()
	}, [initScanner])

	useFocusEffect(
		useCallback(() => {
			setIsScannerVisible(true)
			return () => {
				setIsScannerVisible(false)
			}
		}, []),
	)

	const toggleScanner = useCallback(async () => {
		if (!isScannerSelected) {
			const status = await checkPermissions(PermissionType.camera, {
				navigate,
				navigateToPermScreenOnProblem: true,
				onSuccess: () => setIsScannerSelected(true),
			})
			if (status === RESULTS.DENIED || status === RESULTS.BLOCKED) {
				return
			}
		}
		setIsScannerSelected(!isScannerSelected)
	}, [isScannerSelected, navigate])

	const handleSwitchQr = useCallback(async () => {
		if (!isScannerSelected) {
			await checkPermissions(PermissionType.camera, {
				navigate,
				navigateToPermScreenOnProblem: true,
				onComplete: () => {
					toggleScanner()
				},
				onSuccess: () => {
					toggleScanner()
				},
			})
		} else {
			toggleScanner()
		}
	}, [isScannerSelected, navigate, toggleScanner])

	return (
		<Layout style={[flex.tiny, { backgroundColor: colors['main-background'] }]}>
			<StatusBar backgroundColor={colors['background-header']} barStyle='light-content' />
			<ScrollView
				style={[
					margin.bottom.medium,
					{
						backgroundColor: colors['main-background'],
					},
				]}
			>
				<ShareContainer
					element={isScannerSelected ? <ScanBody visible={isScannerVisible} /> : <ShareQr />}
				>
					<ButtonSettingRow
						state={[
							{
								displayComponent: isScannerSelected && <QrCode size={80} />,
								name: t('settings.share.tap-to-scan'),
								icon: 'camera-outline',
								color: colors['background-header'],
								style: [margin.right.scale(20), height(120), width(120)],
								onPress: handleSwitchQr,
							},
							{
								name: t('settings.share.invite'),
								icon: 'link-outline',
								color: colors['background-header'],
								style: [height(120), height(120), width(120)],
								onPress: async () => {
									if (url) {
										try {
											if (Platform.OS === 'web') {
												Clipboard.setString(url)
											} else {
												await Share.share({ url, message: url })
											}
										} catch (e) {
											console.error(e)
										}
									}
								},
							},
						]}
					/>
				</ShareContainer>
				<View style={[margin.horizontal.medium]}>
					<ButtonSetting
						name={t('settings.share.create-group')}
						onPress={() => navigate('Chat.CreateGroupAddMembers')}
					/>
				</View>
			</ScrollView>
		</Layout>
	)
}
