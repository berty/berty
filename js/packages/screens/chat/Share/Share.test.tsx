import { renderScreen } from '@berty/utils/testing/renderScreen.test'

import { ShareModal } from './Share'

test('Chat.Share renders correctly', () => {
	const { toJSON } = renderScreen('Chat.Share', ShareModal)
	expect(toJSON()).toMatchSnapshot()
})
