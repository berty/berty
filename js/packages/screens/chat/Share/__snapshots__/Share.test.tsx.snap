// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat.Share renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      appearance="default"
      level="1"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ],
        ]
      }
    >
      <RCTScrollView
        style={
          Array [
            Object {
              "marginBottom": 16,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "borderBottomLeftRadius": 40,
                  "borderBottomRightRadius": 40,
                },
                Object {
                  "paddingBottom": 40,
                },
                Object {
                  "paddingTop": 16,
                },
                Object {
                  "backgroundColor": "#535AE4",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "borderRadius": 16,
                  },
                  Object {
                    "marginBottom": 32,
                  },
                  Object {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "backgroundColor": "#FFFFFF",
                    "height": 535,
                    "width": 535,
                  },
                ]
              }
            >
              <View>
                <View
                  style={
                    Array [
                      Object {
                        "marginTop": 32,
                      },
                      Object {
                        "marginBottom": 9,
                      },
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "marginRight": 9,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "zIndex": -1,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            undefined,
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "#FFFFFF",
                              "borderRadius": 12,
                              "height": 24,
                              "justifyContent": "center",
                              "width": 24,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "justifyContent": "center",
                            }
                          }
                        >
                          <View
                            style={
                              Array [
                                undefined,
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#b8bd4a",
                                  "borderRadius": 12,
                                  "height": 24,
                                  "justifyContent": "center",
                                  "width": 24,
                                },
                              ]
                            }
                          >
                            <Text
                              adjustsFontSizeToFit={true}
                              style={
                                Array [
                                  Object {
                                    "fontSize": 30,
                                  },
                                  Object {
                                    "color": "#383B62",
                                    "fontFamily": "Open Sans",
                                  },
                                  Object {
                                    "color": "#FFFFFF",
                                    "fontSize": 12,
                                    "includeFontPadding": false,
                                    "paddingBottom": 1.2,
                                    "textAlign": "center",
                                  },
                                ]
                              }
                            >
                              ?
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#383B62",
                          "fontFamily": "Open Sans",
                        },
                        undefined,
                      ]
                    }
                  >
                    
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "height": 337.5,
                      "justifyContent": "center",
                      "width": 337.5,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#FFFFFF",
                        "height": "100%",
                        "justifyContent": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Image
                      source="AssetMock"
                      style={
                        Object {
                          "height": "40%",
                          "maxHeight": 80,
                          "maxWidth": 170,
                          "width": "80%",
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <RCTScrollView
              contentContainerStyle={
                Array [
                  Object {
                    "flex": 1,
                  },
                  Object {
                    "alignItems": "stretch",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  },
                  Object {
                    "marginTop": 16,
                  },
                  Object {
                    "paddingBottom": 9,
                  },
                  null,
                  Object {
                    "justifyContent": "center",
                  },
                ]
              }
              horizontal={true}
              scrollEnabled={false}
              showsHorizontalScrollIndicator={false}
            >
              <View>
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 16,
                      "height": 120,
                      "justifyContent": "center",
                      "marginRight": 20,
                      "opacity": 1,
                      "padding": 16,
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": 0,
                      },
                      "shadowOpacity": 0.2,
                      "shadowRadius": 2,
                      "width": 120,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                        ],
                      }
                    }
                  >
                    <RNSVGMock
                      fill="#535AE4"
                      height={30}
                      viewBox="0 0 24 24"
                      width={30}
                    />
                  </View>
                  <Text
                    numberOfLines={2}
                    style={
                      Array [
                        Object {
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#383B62",
                          "fontFamily": "Open Sans",
                        },
                        Array [
                          Object {
                            "textAlign": "center",
                          },
                          Object {
                            "fontSize": 30,
                          },
                          null,
                          Object {
                            "paddingTop": 6,
                          },
                        ],
                      ]
                    }
                  >
                    settings.share.tap-to-scan
                  </Text>
                </View>
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 16,
                      "height": 120,
                      "justifyContent": "center",
                      "opacity": 1,
                      "padding": 16,
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": 0,
                      },
                      "shadowOpacity": 0.2,
                      "shadowRadius": 2,
                      "width": 120,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                        ],
                      }
                    }
                  >
                    <RNSVGMock
                      fill="#535AE4"
                      height={30}
                      viewBox="0 0 24 24"
                      width={30}
                    />
                  </View>
                  <Text
                    numberOfLines={2}
                    style={
                      Array [
                        Object {
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#383B62",
                          "fontFamily": "Open Sans",
                        },
                        Array [
                          Object {
                            "textAlign": "center",
                          },
                          Object {
                            "fontSize": 30,
                          },
                          null,
                          Object {
                            "paddingTop": 6,
                          },
                        ],
                      ]
                    }
                  >
                    settings.share.invite
                  </Text>
                </View>
              </View>
            </RCTScrollView>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 16,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "shadowColor": "#000000",
                  },
                  Object {
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.2,
                    "shadowRadius": 2,
                  },
                  Object {
                    "borderRadius": 14,
                    "marginTop": 20,
                  },
                ]
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 48,
                    "justifyContent": "space-between",
                    "opacity": 1,
                    "padding": 12,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "width": "75%",
                    }
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#383B62",
                          "fontFamily": "Open Sans",
                        },
                        Array [
                          Object {
                            "marginLeft": 9,
                          },
                        ],
                      ]
                    }
                  >
                    settings.share.create-group
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "transform": Array [
                        Object {
                          "scale": 1,
                        },
                      ],
                    }
                  }
                >
                  <RNSVGMock
                    fill="#393C63"
                    height={20}
                    viewBox="0 0 24 24"
                    width={20}
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <TextInput
                allowFontScaling={true}
                onChangeText={[Function]}
                placeholder="Paste link here"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flex": 1,
                  }
                }
                underlineColorAndroid="transparent"
              />
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      undefined,
                    ]
                  }
                >
                  Confirm
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
